src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/ActivitySource.cs

```diff
@@ -0,0 +1,357 @@
+        }
+
+        /// <summary>
+        /// Add a listener to the <see cref="Activity"/> starting and stopping events.
+        /// </summary>
+        /// <param name="listener"> The <see cref="ActivityListener"/> object to use for listeneing to the <see cref="Activity"/> events.</param>
+        public static void AddActivityListener(ActivityListener listener)
+        {
+            if (listener == null)
+            {
+                throw new ArgumentNullException(nameof(listener));
+            }
+
+            if (s_allListeners.AddIfNotExist(listener))
+            {
+                s_activeSources.EnumWithAction(source => {
+                    var shouldListenTo = listener.ShouldListenTo;
+                    if (shouldListenTo != null && shouldListenTo(source))
+                    {
+                        source.AddListener(listener);
+                    }
+                });
+            }
+        }
+
+        internal void AddListener(ActivityListener listener)
+        {
+            if (_listeners == null)
+            {
+                Interlocked.CompareExchange(ref _listeners, new SynchronizedList<ActivityListener>(), null);
+            }
+
+            _listeners.AddIfNotExist(listener);
+        }
+
+        internal static void DetachListener(ActivityListener listener)
+        {
+            s_allListeners.Remove(listener);
+
+            s_activeSources.EnumWithAction(source => {
+                var listeners = source._listeners;
+                listeners?.Remove(listener);
+            });
+        }
+
+        internal void NotifyActivityStart(Activity activity)
+        {
+            Debug.Assert(activity != null);
+
+            // _listeners can get assigned to null in Dispose.
+            SynchronizedList<ActivityListener>? listeners = _listeners;
+            if (listeners != null &&  listeners.Count > 0)
+            {
+                listeners.EnumWithAction(listener => listener.ActivityStarted?.Invoke(activity));
+            }
+        }
+
+        internal void NotifyActivityStop(Activity activity)
+        {
+            Debug.Assert(activity != null);
+
+            // _listeners can get assigned to null in Dispose.
+            SynchronizedList<ActivityListener>? listeners = _listeners;
+            if (listeners != null &&  listeners.Count > 0)
+            {
+                listeners.EnumWithAction(listener => {
+                    listeners.EnumWithAction(listener => listener.ActivityStopped?.Invoke(activity));
+                });
+            }
+        }
+    }
+
+    // SynchronizedList<T> is a helper collection which ensure thread safety on the collection
+    // and allow enumerating the collection items and execute some action on the enumerated item and can detect any change in the collection
+    // during the enumeration which force restarting the enumeration again.
+    // Causion: We can have the action executed on the same item more than once which is ok in our scenarios.
```

Comment:
```suggestion
    // Caution: We can have the action executed on the same item more than once which is ok in our scenarios.
```

Reply:
why English lack having `Causion` :-) interesting the editor spell checker didn't complain. of course, I'll fix it. thanks.